src/machine/shared_library.rs

```diff
@@ -0,0 +1,269 @@
+        use serde_json::{json, Value};
+
+        /// Create a new instance of the Scryer Machine.
+        ///
+        ///
+        /// It is the caller's responsibility to maintain a reference to the machine pointer
+        /// created by this function, passing it to [`machine_free`] to deallocate
+        /// resources. Failure to do so may lead to memory leaks.
+        ///
+        #[export_name = "scryer_machine_new"]
+        pub extern "C" fn machine_new() -> *mut Machine {
+            let machine = Box::into_raw(Box::new(Machine::new_lib()));
+            machine
+        }
+
+        /// Frees the memory occupied by a [`Machine`] object.
+        ///
+        /// # Safety
+        ///
+        /// * It is the caller's responsibility to ensure that the input is a valid [`Machine`] pointer
+        /// created by [`machine_new`] that has not yet been freed by this function.
+        /// * Any [`QueryState`] instances associated with this machine should be deallocated prior
+        /// to calling this function on the input, otherwise those [`QueryState`] instances may
+        /// not be deallocated until the program terminates.
+        ///
+        #[export_name = "scryer_machine_free"]
+        pub unsafe extern "C" fn machine_free(ptr: *mut Machine) {
+            unsafe {
+                drop(Box::from_raw(ptr));
+            }
+        }
+
+        /// Returns a new query generator for the given virtual machine.
+        ///
+        /// # Safety
+        ///
+        /// Caller must satisfy the following preconditions for this function:
+        ///
+        /// * Valid [`Machine`] pointer created with [`machine_new`] that has not yet been freed.
+        /// * the input fulfill the safety requirements of [`CStr::From_ptr`].
+        /// * There must be no other [`QueryState`] for this [`Machine`] started by
+        /// [`run_query_iter`] that has not yet been freed with
+        /// [`query_state_free`] or the [`Machine`] state will enter an undefined
+        /// configuration with unpredictable results.
+        ///
+        /// Other concerns:
+        /// * after invoking this function, calling any other function besides [`run_query_next`]
+        /// before invoking [`query_state_free`] on the [`Machine`] pointer and [`QueryState`] pointer
+        /// will leave the [`Machine`] in an undefined state
+        ///
+        #[export_name = "scryer_run_query_iter"]
+        pub unsafe extern "C" fn run_query_iter(
+            machine: &mut Machine,
+            input: *const c_char,
+        ) -> *mut QueryState {
+            match unsafe { CStr::from_ptr(input) }.to_str() {
+                Ok(input) => {
+                    let string = input.to_string();
+                    let query_state = machine.run_query_iter(string);
+                    Box::into_raw(Box::new(query_state))
+                }
+                Err(err) => {
+                    eprintln!("Query input wasn't valid UTF-8: {err:?}");
```

Comment:
~This error should probably be returned some way, like in an out pointer, so that it can be handled.~

Actually, this seems to be the only error possible in this case here, so maybe remove this `eprintln!()` and mention in the documentation that if the string given isn't valid UTF-8 then a null pointer is returned. Even if you want to keep this, you should probably use the `Display` implementation of the error instead of the `Debug`:

```rust
eprintln!("Query input wasn't valid UTF-8: {err}");
```